src/python/pants/backend/jvm/tasks/jvm_compile/jvm_compile.py

```diff
@@ -428,55 +432,53 @@ def execute(self):
 
     relevant_targets = list(self.context.targets(predicate=self.select))
 
-    # If we are only exporting jars then we can omit some targets from the runtime_classpath.
-    if (
-        self.context.products.is_required_data("export_dep_as_jar_classpath") and
-        not self.context.products.is_required_data("runtime_classpath")
-    ):
-      # Filter modulized targets from invalid targets list.
-      target_root_addresses = [t.address for t in set(self.context.target_roots)]
-      dependees_of_target_roots = self.context.build_graph.transitive_dependees_of_addresses(target_root_addresses)
-      relevant_targets = list(set(relevant_targets) - dependees_of_target_roots)
-
-    if not relevant_targets:
-      return
-
     # Clone the compile_classpath to the runtime_classpath.
     classpath_product = self.create_classpath_product()
 
     fingerprint_strategy = DependencyContext.global_instance().create_fingerprint_strategy(
         classpath_product)
-    # Note, JVM targets are validated (`vts.update()`) as they succeed.  As a result,
-    # we begin writing artifacts out to the cache immediately instead of waiting for
-    # all targets to finish.
-    with self.invalidated(relevant_targets,
-                          invalidate_dependents=True,
-                          fingerprint_strategy=fingerprint_strategy,
-                          topological_order=True) as invalidation_check:
-
-      compile_contexts = {vt.target: self.create_compile_context(vt.target, vt.results_dir)
-                          for vt in invalidation_check.all_vts}
-
-      self.do_compile(
-        invalidation_check,
-        compile_contexts,
-        classpath_product,
+
+    dependees_of_target_roots = None
+    # If we are only exporting jars then we can omit some targets from the runtime_classpath.
+    if self.context.products.is_required_data("export_dep_as_jar_signal"):
+      # Filter modulized targets from invalid targets list.
+      target_roots_in_play = set(relevant_targets) & set(self.context.target_roots)
+      addresses_in_play = [t.address for t in target_roots_in_play]
+      dependees_of_target_roots = set(
```

Comment:
I feel like this is reimplementing part of the concept of "modulizable targets". Ideally, I'd like that logic (is this target modulizable) to be implemented in a single place.

However, I think this is out of scope of this PR, and I should have caught it in your earlier PR.

Reply:
Yeah I agree that would be nice, but practically we have had two implementations before this PR. One in export_dep_as_jar.py and one in jvm_compile.py. I think I'd like to deal with that in a follow up because there are subtle differences between the two implementations. export-dep-as-jar controls what really ends up as a target vs. a library. These changes are about safely controlling which targets we can create compiler output products for.

Reply:
Sounds good. Thanks!